storage/signed_urls/generate_signed_urls.py

```diff
@@ -0,0 +1,159 @@
+# Copyright 2018 Google, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+
+"""This application demonstrates how to construct a Signed URL for objects in
+   Google Cloud Storage.
+
+For more information, see the README.md under /storage and the documentation
+at https://cloud.google.com/storage/docs/access-control/signing-urls-manually.
+"""
+
+# [START storage_signed_url_all]
+# [START storage_signed_url_dependencies]
+import collections
+import datetime
+import hashlib
+import sys
+
+# pip install six
+from six.moves.urllib.parse import quote
+
+# pip install google-auth
+from google.oauth2 import service_account
+# [END storage_signed_url_dependencies]
+
+
+def generate_signed_url(service_account_file, bucket_name, object_name,
+                        expiration, http_method='GET', query_parameters=None,
+                        headers=None):
+
+    if expiration > 604800:
+        print('Expiration Time can\'t be longer than 604800 seconds (7 days).')
+        sys.exit(1)
+
+    # [START storage_signed_url_canonical_uri]
+    escaped_object_name = quote(object_name, safe='')
+    canonical_uri = "/{}/{}".format(bucket_name, escaped_object_name)
```

Comment:
Prefer single-quote strings over double-quotes, there are a few more below to get too